build: add `-Wundef`

fanquake opened the pull request:
Turn on `-Wundef`.

[> Warn if an undefined identifier is evaluated in an #if directive. Such identifiers are replaced with zero.](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wundef).

Note that this is still beneficial with CMake, and may even be nice to have enabled prior, to catch any change in behaviour.

If we end up with this enabled, it should probably be enough to fix #16419.

fanquake committed:
build: add -Wundef

"Warn if an undefined identifier is evaluated in an #if directive. Such
identifiers are replaced with zero."

fanquake committed:
zmq: use #ifdef ENABLE_ZMQ

fanquake committed:
randomenv: use ifdef over if

randomenv.cpp:48:5: warning: 'HAVE_VM_VM_PARAM_H' is not defined, evaluates to 0 [-Wundef]

randomenv.cpp:51:5: warning: 'HAVE_SYS_RESOURCES_H' is not defined, evaluates to 0 [-Wundef]

randomenv.cpp:424:5: error: 'HAVE_SYSCTL' is not defined, evaluates to 0 [-Werror,-Wundef]

laanwj commented:
Concept ACK, issues with misspelled or otherwise missing defines can be really sneaky (and have caused some hard to find bugs in the past), this ia good check to have.

laanwj commented on src/randomenv.cpp:
```diff
@@ -44,15 +44,15 @@
 #include <ifaddrs.h>
 #endif
-#if HAVE_SYSCTL
+#ifdef HAVE_SYSCTL
```
Would it be possible to change these to values instead of defined/undefined? It seems to me that changing these to `ifdef` makes it less safe because you lose the protection added by the warning in the first place. (say, when you forget to include bitcoin-config)

fanquake replied:
> Would it be possible to change these to values 0/1 instead of defined/undefined?

These are currently checked via: `AC_CHECK_HEADERS([sys/select.h sys/prctl.h sys/sysctl.h vm/vm_param.h sys/vmmeter.h sys/resources.h])`. If we wanted, to we could change this.

laanwj replied:
i still think it'd be better, but probably not worth doing this last-minute change for autotools.
Don't know how the CMake port handles these kind of checks anyway, but might consider it there.

fanquake replied:
> but probably not worth doing this last-minute change for autotools.

Not sure. Everything not done for Autotools, but done in CMake is just more bebaviour/divergence to maintain, so I'd prefer to do as much as possible prior.

laanwj reviewed (approved):
ACK cc09c165eb0bbb4b41a9132734bee55969e9d9a8

hebasto reviewed (commented):
Concept ACK.

hebasto reviewed (approved):
ACK cc09c165eb0bbb4b41a9132734bee55969e9d9a8.

I suggest to consider updating a help string in https://github.com/bitcoin/bitcoin/blob/a46065e36cf868265c909dc5edf29dc17be53c1f/configure.ac#L1133 to "Define to 1 if O_CLOEXEC flag is available, and to 0 if not."

As a follow up, the https://github.com/bitcoin/bitcoin/pull/16344 might be reconsidered.

fanquake commented:
Rebased on #25972. Please review that first.

> I suggest to consider updating a help string in 

I don't touch that here, and it's not clear why we need to change a single instsance out of many help strings that could be similarly updated, so going to leave things as they are.

theuni commented:
Any reason this is still in draft? I agree it seems useful do enable before the CMake switch.

fanquake committed:
refactor: use #ifdef HAVE_SOCKADDR_UN

```bash
init.cpp:526:5: error: "HAVE_SOCKADDR_UN" is not defined, evaluates to 0 [-Werror=undef]
  526 | #if HAVE_SOCKADDR_UN
      |     ^~~~~~~~~~~~~~~~
init.cpp:541:5: error: "HAVE_SOCKADDR_UN" is not defined, evaluates to 0 [-Werror=undef]
  541 | #if HAVE_SOCKADDR_UN
      |     ^~~~~~~~~~~~~~~~
init.cpp:1318:5: error: "HAVE_SOCKADDR_UN" is not defined, evaluates to 0 [-Werror=undef]
 1318 | #if HAVE_SOCKADDR_UN
```
```
netbase.cpp:26:5: error: "HAVE_SOCKADDR_UN" is not defined, evaluates to 0 [-Werror=undef]
   26 | #if HAVE_SOCKADDR_UN
      |     ^~~~~~~~~~~~~~~~
netbase.cpp:221:5: error: "HAVE_SOCKADDR_UN" is not defined, evaluates to 0 [-Werror=undef]
  221 | #if HAVE_SOCKADDR_UN
      |     ^~~~~~~~~~~~~~~~
netbase.cpp:496:5: error: "HAVE_SOCKADDR_UN" is not defined, evaluates to 0 [-Werror=undef]
  496 | #if HAVE_SOCKADDR_UN
      |     ^~~~~~~~~~~~~~~~
netbase.cpp:531:5: error: "HAVE_SOCKADDR_UN" is not defined, evaluates to 0 [-Werror=undef]
  531 | #if HAVE_SOCKADDR_UN
      |     ^~~~~~~~~~~~~~~~
netbase.cpp:639:5: error: "HAVE_SOCKADDR_UN" is not defined, evaluates to 0 [-Werror=undef]
  639 | #if HAVE_SOCKADDR_UN
```

hebasto commented:
> This was the multiprocess job failure:

cc @ryanofsky

hebasto commented:
> We'll probably need to fixup the failing ARM and previous releases job upstream:
> 
> ```shell
> In file included from minisketch/src/fields/generic_common_impl.h:12,
>                  from minisketch/src/fields/generic_4bytes.cpp:12:
> minisketch/src/fields/../int_utils.h:162:7: error: "HAVE_CLZ" is not defined, evaluates to 0 [-Werror=undef]
>   162 | #elif HAVE_CLZ
>       |       ^~~~~~~~
> ```

The upstream issue has been addressed in https://github.com/sipa/minisketch/pull/88.

ryanofsky commented:
Hmm, this is the first time I've seen the -Wundef warning, and my initial reaction is to be skeptical of it, because it seems like it would require patching a lot of upstream headers, and I'm not sure if it it just forces writing `#if defined(SYMBOL)` instead of `#if SYMBOL` everywhere it is likely to prevent any bugs. It seems more like it is breaking a commonly used idiom.

But if I'm not seeing the benefits, or others just think Wundef is more useful, I can create a patch for depends and an upstream pull request for capnproto. I don't want capnproto to get in the way of having better diagnostics or improvements to the build system.

ryanofsky committed:
build: Suppress warnings from boost and capnproto in multiprocess code

Without this change there are errors from boost like:

/ci_container_base/depends/i686-pc-linux-gnu/include/boost/signals2/expired_slot.hpp:23:28: error: 'what' overrides a member function but is not marked 'override' [-Werror,-Wsuggest-override]
/ci_container_base/depends/i686-pc-linux-gnu/include/boost/signals2/detail/signal_template.hpp:750:32: error: 'lock_pimpl' overrides a member function but is not marked 'override' [-Werror,-Wsuggest-override]
/ci_container_base/depends/i686-pc-linux-gnu/include/boost/signals2/connection.hpp:150:22: error: 'connected' overrides a member function but is not marked 'override' [-Werror,-Wsuggest-override]

There do not seem to be errors from capnproto currently, but add a suppression
for it, too, to be consistent with other libraries.

ryanofsky commented:
> Hmm, this is the first time I've seen the -Wundef warning, and my initial reaction is to be skeptical of it, because it seems like it would require patching a lot of upstream headers

Actually I think I'm wrong about this. It seems like compiler will not trigger warnings in headers if headers are found on an -isystem path instead of a normal include path. So as long as we include capnproto headers with -isystem, maybe there will be no problems. I actually implemented that last week in commit a491cea18ee63e514359a5ba699e8cc159664efc from #10102 to suppress warnings from boost headers, and just realized that fix could be applicable here.

Maybe a491cea18ee63e514359a5ba699e8cc159664efc could be cherry-picked here to fix the multiprocess failure https://github.com/bitcoin/bitcoin/pull/29876#issuecomment-2157739910?

fanquake commented:
> Maybe https://github.com/bitcoin/bitcoin/commit/a491cea18ee63e514359a5ba699e8cc159664efc could be cherry-picked here to fix the multiprocess failure https://github.com/bitcoin/bitcoin/pull/29876#issuecomment-2157739910?

That very-much looks like the correct fix.

fanquake commented:
This is ready for further review.

ryanofsky reviewed (approved):
Code review ACK 9e5bd977688f28a29806236f0faa55d5272f5b65

fanquake commented:
Rebased to fix conflicts.

hebasto reviewed (approved):
ACK e3dc64f4990a15df3fd6147831f66fc2a31c71ad, I have reviewed the code and it looks OK.

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/264.
